site/essays/Git Workflow for Managing Long-Running Upgrades A.md

```diff
@@ -0,0 +1,345 @@
+    </picture>
+    <figcaption>fixes applied to separate branches</figcaption>
+</figure>
+
+Each of these will merge in its own time, after being reviewed and passing tests on CI. Once the commits are merged, I’ll update to the current `master` on `new-horizons-alt`, just as before:
+
+```
+$ git checkout master
+$ git pull
+```
+
+Now `master` contains all of the changes I made, though not necessarily in the same order they were in the original upgrade branch that I cherry-picked them from—but that’s fine. After all, that’s exactly why they were broken out into discrete commits! The commit graph might end up being the *previous* `HEAD` on `master`, then `C`, then `A`, then `B`—the order will just be whatever order they happened to land it:
+
+<figure>
+    <picture>
+        <source srcset="https://cdn.chriskrycho.com/file/chriskrycho-com/images/essays/git-workflow/multiple-commits-master-light.png" media="(prefers-color-scheme: light)">
+        <source srcset="https://cdn.chriskrycho.com/file/chriskrycho-com/images/essays/git-workflow/multiple-commits-master-dark.png" media="(prefers-color-scheme: dark)">
+        <img src="https://cdn.chriskrycho.com/file/chriskrycho-com/images/essays/git-workflow/multiple-commits-master-light.png">
+    </picture>
+    <figcaption>multiple commits on <code>master</code></figcaption>
+</figure>
+
+At this point, `master` contains all the fixes I made on `pluto` in the `new-horizons` branch. I can now rebase so once again my working copy *only* contains the `upgrade` commit on top of `master`.
+
+```
+$ git pull --rebase origin master
+```
+
+Now, my commit graph for `pluto` is once again just `master` with one extra commit, the upgrade, all sitting on top of the changes I made in `A`, `B`, and `C`:
+
+<figure>
+    <picture>
+        <source srcset="https://cdn.chriskrycho.com/file/chriskrycho-com/images/essays/git-workflow/fully-merged-light.png" media="(prefers-color-scheme: light)">
+        <source srcset="https://cdn.chriskrycho.com/file/chriskrycho-com/images/essays/git-workflow/fully-merged-dark.png" media="(prefers-color-scheme: dark)">
+        <img src="https://cdn.chriskrycho.com/file/chriskrycho-com/images/essays/git-workflow/fully-merged-light.png">
+    </picture>
+    <figcaption>after rebasing with <code>A</code>, <code>B</code>, and <code>C</code>.</figcaption>
+</figure>
+
+Finally, I clean up the branches I created for the fixes.
+
+```sh
+# the long form
+$ git branch --delete fix-a fix-b fix-c
+# the short form
+$ git branch -d fix-a fix-b fix-c
+```
+
+## Summary
+
+When working with long-running projects (like difficult dependency upgrades), it’s helpful to land as many changes as possible on the main development branch instead of landing them all at once in a big bang. In at least some cases, we can accomplish this by making good use of the capabilities of distributed version control systems (<abbr>DVCS</abbr>s) like Git: clones, remotes, cherry-picking, and branches.
```

Comment:
This worked better as an introduction. I should move it back up there and find some *other* way to wrap up the post: "For anyone else out there working on long-running and difficult tasks like upgrades…"

Reply:
Fixed in be16756.